### ISSUE TITLE
About the tag 'EQ'

### ISSUE THREAD

**Issue by Pavansid94** [lixin4ever/BERT-E2E-ABSA#40]
First of all, great work.

I wanted to do a small Inquiry about the tag 'EQ' in all the tagging schemes. May I know what that actually means? I didn't see anything related to it in the papers mentioned in the README. I also see that in the [`ot2bieos_ts()`](https://github.com/lixin4ever/BERT-E2E-ABSA/blob/master/seq_utils.py#L19) in the `seq_utils.py` the tag 'EQ' is replaced with an 'O'. Could you enlighten me on why this is?

**Comment by lixin4ever** [comment]
Thank you for your attention.

As we all know one word is likely to be segmented into multiple subwords after applying BERT tokenization. To better accommodate the subword-level predictions, we keep the tag of the head subword (i.e., the first subword) identical to the word tag and set the tags for the subsequent non-head subwords as 'EQ' (which means 'equivalent' to the head subword). During training, the losses on the 'EQ' tag are considered (NOT ELEGANT!! CAN BE AVOIDED by adding a mask), however,  in the inference stage, we only pick the predictions of the head subwords as the output word tags (i.e., ts_tag_sequence in ot2bieos_ts). Since we model the 'EQ' during training, it is possible (but of a very low chance) that the 'EQ' will appear in the output word tags. In case of the appearance of 'EQ' tags, we simply replace the 'EQ' in the (word) tag sequence with 'O'.

**Comment by Pavansid94** [comment]
Thanks, I appreciate your quick response.

So, In other words, it is advisable to mask the non-head subwords and remove them altogether for training. Is this right? are there any instances where this could be a problem? But also, you say 

> To better accommodate the subword-level predictions...

is this only to know that there are indeed non-head subwords that were suggested by the BERT tokenizer and we need a way to probably recognize them OR are there any other reasons as well?

Also, I had another question unrelated to the main query. After reading through your paper (great read, again), I see that the pretrained 'bert-base-uncased' was made use of. Do you see any reason where probably other BERT variants (or its siblings such as RoBERTa or distilBERT) could lead to better result (if at all)? What do you say?

**Comment by lixin4ever** [comment]
Q1 -- You should not mask the non-head subwords (because they are necessary for calculating the accurate contextualized representations) but ignore the losses from these positions.

Q2 -- This project had been started (in early 2019) before the release date of RoBERTa and distillBERT. One thing for sure is that you will obtain better results if you change the pretrained text encoder from BERT to RoBERTa.

**Comment by Pavansid94** [comment]
I see. cool. 
Thanks for the replies, once again. really appreciate it!
